packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,378 @@
+        whenDefined: nativeWhenDefined,
+        get: nativeGet,
+    } = window.customElements;
+
+    // patch for the global registry define mechanism
+    CustomElementRegistry.prototype.define = function define(
+        this: CustomElementRegistry,
+        tagName: string,
+        constructor: CustomElementConstructor,
+        options?: ElementDefinitionOptions
+    ): void {
+        if (options && options.extends) {
+            throw new DOMException('NotSupportedError: ');
+        }
+        nativeGet.call(this, tagName); // SyntaxError if The provided name is not a valid custom element name.
+        tagName = tagName.toLowerCase();
+        if (_definitionsByTag.get(tagName) !== undefined) {
+            throw new DOMException(
+                `Failed to execute 'define' on 'CustomElementRegistry': the name "${tagName}" has already been used with this registry`
+            );
+        }
+        const definition = getDefinitionForConstructor(constructor);
+        if (_definitionsByClass.get(constructor) !== undefined) {
+            throw new DOMException(
+                `Failed to execute 'define' on 'CustomElementRegistry': this constructor has already been used with this registry`
+            );
+        }
+        _definitionsByTag.set(tagName, definition);
+        _definitionsByClass.set(constructor, definition);
+        let PivotCtor = _pivotCtorByTag.get(tagName);
+        if (!PivotCtor) {
+            PivotCtor = createPivotingClass(tagName, definition);
+            // Register a pivoting class which will handle global registry initializations
+            nativeDefine.call(nativeRegistry, tagName, PivotCtor);
+        }
+        // For globally defined custom elements, the definition associated
+        // to the UserCtor has a back-pointer to PivotCtor in case the user
+        // new the UserCtor, so we know how to create the underlying element.
+        definition.PivotCtor = PivotCtor;
+        // Upgrade any elements created in this scope before define was called
+        // which should be exhibit by LWC using a tagName (in a template)
+        // before the same tagName is registered as a global, while others
+        // are already created and waiting in the global context, that will
+        // require immediate upgrade when the new global tagName is defined.
+        const awaiting = _awaitingUpgrade.get(tagName);
+        if (awaiting) {
+            _awaitingUpgrade.delete(tagName);
+            for (const element of awaiting) {
+                const registeredDefinition = pendingRegistryForElement.get(element);
+                if (registeredDefinition) {
+                    pendingRegistryForElement.delete(element);
+                    internalUpgrade(element, registeredDefinition, definition);
+                }
+            }
+        }
+        // Flush whenDefined callbacks
+        const resolver = _definedResolvers.get(tagName);
+        if (resolver) {
+            resolver(constructor);
```

Comment:
it seems to me that, at this point, we should clear the promises/resolvers from the `_definedPromises` and `_definedResolvers` maps. Otherwise it could cause a memory leak (if I grok the code correctly).